### DISCUSSION TITLE
Confusion with `ArrayOptions.contains`

### DISCUSSION THREAD

**Discussion by thousynth** [sinclairzx81/typebox#607] [Q&A] [answered]
I've two use cases here:
* Only one option object must have `isCorrect: true` for Multiple Choice Question for single answers
* More than one option  objects must have `isCorrect: true` for Multiple Choice Question for multiple answers
the flag `isCorrect` for at least one option object in an array has to be true.
* For each question there must be at least one correct option.

I've two Schemas one for a Question and one for Option.
The OptionsDTO defines the option's schema and the QuestionsDTO has an array of options inside them. 
I'm trying to create such schema using the `contains` option in the `ArrayOptions`. 
It throws a status code `500` with `undefined is not an object (evaluating 'value.options.reduce')` message.

What would be the correct way to achieve this?

```js
const OptionsDTO = t.Object({
  text: t.String(),
  isCorrect: t.Boolean()
})

const QuestionDTO = t.Object({
  text: t.String(),
  options: t.Array(OptionsDTO, {
    minItems: 2,
    maxItems: 5,
    contains: t.Object({
      text: t.String(),
      isCorrect: t.Literal(true)
    })
  })
})
```

**Comment by sinclairzx81** [comment]
@cheersmas Heya, have you tried `minContains` and `maxContains` ?

```typescript
import { Type } from '@sinclair/typebox'
import { Value } from '@sinclair/typebox/value'

const Answer = Type.Object({
  text: Type.String(),
  isCorrect: Type.Boolean()
})
const SingleChoice = Type.Object({
  text: Type.String(),
  options: Type.Array(Answer, {
    minContains: 1, // expect ONLY 1 element to contain schematic
    maxContains: 1,
    contains: Type.Object({
      text: Type.String(),
      isCorrect: Type.Literal(true)
    })
  })
})
const MultiChoice = Type.Object({
  text: Type.String(),
  options: Type.Array(Answer, {
    minContains: 1, // expect at least 1 element to contain schematic
    contains: Type.Object({
      text: Type.String(),
      isCorrect: Type.Literal(true)
    })
  })
})
// ------------------------------------------------------------------
// SingleChoice
// ------------------------------------------------------------------
const A = Value.Check(SingleChoice, { // true
  text: 'SingleChoice',
  options: [
    { text: 'A', isCorrect: true },
    { text: 'B', isCorrect: false }
  ]
})
const B = Value.Check(SingleChoice, { // false
  text: 'SingleChoice',
  options: [
    { text: 'A', isCorrect: true },
    { text: 'B', isCorrect: true }
  ]
})
// ------------------------------------------------------------------
// MultiChoice
// ------------------------------------------------------------------
const C = Value.Check(MultiChoice, { // true
  text: 'MultiChoice',
  options: [
    { text: 'A', isCorrect: true },
    { text: 'B', isCorrect: true }
  ]
})
const D = Value.Check(MultiChoice, { // false
  text: 'MultiChoice',
  options: [
    { text: 'A', isCorrect: false },
    { text: 'B', isCorrect: false }
  ]
})
```

**Comment by thousynth** [comment]
I tried this, although your answer validates correctly. It isn't working for me when I use it to validate my body in a `POST` request in `Elysia`. I'm using `bun`.
It throws the same error. If I try to debug this, on turning on caught exceptions it lands me here. The value here is:
<img width="1244" alt="image" src="https://github.com/sinclairzx81/typebox/assets/41203791/ebaa4db3-0c20-473f-a93b-3ae7e8d944a5">

You see the value is an array and it tries to get the `options` key from it which is `undefined`:
<img width="768" alt="image" src="https://github.com/sinclairzx81/typebox/assets/41203791/fba101ef-47e5-446c-abbb-13612fd9e064">


Here is my complete server code.

```js
import { Elysia, t } from 'elysia'
import mongoose from 'mongoose'
import { Quiz } from './models/Quiz'

const PORT = 8000
const MONGODB_PATH = 'mongodb://localhost:27017/quiz_db'

const OptionDTO = t.Object({
  text: t.String(),
  isCorrect: t.Boolean()
})

const QuestionDTO = t.Object({
  text: t.String(),
  options: t.Array(OptionDTO, {
    minContains: 1,
    maxContains: 1,
    // minItems: 2,
    // maxItems: 5,
    contains: t.Object({
      text: t.String(),
      isCorrect: t.Literal(true)
    })
  })
})

const QuizDTO = t.Object({
  text: t.String(),
  questions: t.Array(QuestionDTO, { minItems: 1, maxItems: 10 })
})

async function main() {
  const app = new Elysia()

  try {
    await mongoose.connect(MONGODB_PATH)
  } catch (err) {
    console.error(err);
  }

  app.get('/', () => 'hello elysia')
  app.post('/api/quiz', async ({ body, ...rest }) => {
    console.log('something')
  }, {
    body: QuizDTO,
  })
  app.get('/api/quiz', async (context) => {
    try {
      const quizzes = await Quiz.find({});
      return quizzes
    } catch (err) {
      return
    }
  })
  app.listen(PORT)

  console.log(`running at: ${app.server?.hostname}:${app.server?.port}`)
}

main().catch(console.error)
```

**Reply by sinclairzx81** [reply]
@cheersmas This looks like a compiler bug as the check function should never throw. The issue appears to be the incorrect `value.options.reduce` (which should just be `value.reduce`), I can publish a patch for this within the next hour.

How easy is it for you to update TypeBox in Elysia?

**Reply by sinclairzx81** [reply, marked as answer]
@cheersmas Hi. I have just published a fix for this which should prevent errors being thrown (this was a good catch)

The update is on `@sinclair/typebox@0.31.17` so you will need to ensure Elysia is on this version. Associated PR can be found https://github.com/sinclairzx81/typebox/pull/608.

**Reply by thousynth** [reply]
Since, it is a minor version. I think a reinstall of ElysiaJS should fix this. I've raised a PR here
https://github.com/elysiajs/elysia/pull/228
